File: packages/flags-flutter/lib/src/user_context.dart
/// User context for evaluating flags.
///
/// Mirrors `SignaKitUserContext` from `packages/flags-node/src/client.ts`.
library;

import 'dart:convert';

import 'bot_patterns.dart';
import 'client.dart';
import 'constants.dart';
import 'decision.dart';
import 'types.dart';

UserAttributes? _sanitizeAttributes(UserAttributes? attributes) {
  if (attributes == null || attributes.isEmpty) return null;

  final out = <String, Object?>{};
  final keys = attributes.keys.take(kMaxAttributesCount);

  for (final key in keys) {
    final sKey = key.length > kMaxAttributeKeyLength
        ? key.substring(0, kMaxAttributeKeyLength)
        : key;
    final value = attributes[key];

    if (value == null) continue;
    if (value is String) {
      out[sKey] = value.length > kMaxAttributeValueLength
          ? value.substring(0, kMaxAttributeValueLength)
          : value;
    } else if (value is List) {
      // Limit list size to 100, truncate string values.
      final limited = value.take(100).map((v) {
        if (v is String && v.length > kMaxAttributeValueLength) {
          return v.substring(0, kMaxAttributeValueLength);
        }
        return v;
      }).toList(growable: false);
      out[sKey] = limited;
    } else {
      out[sKey] = value;
    }
  }

  return out;
}

String _truncate(String value, int max) {
  if (value.length <= max) return value;
  return value.substring(0, max);
}

class SignaKitUserContext {
  SignaKitUserContext({
    required SignaKitClient client,
    required this.userId,
    UserAttributes attributes = const <String, Object?>{},
  })  : _client = client,
        _isBot = isBot(attributes[r'$userAgent'] as String?),
        // Strip $userAgent — not used for targeting.
        attributes = Map<String, Object?>.from(attributes)..remove(r'$userAgent');

  final SignaKitClient _client;
  final String userId;
  final UserAttributes attributes;
  final bool _isBot;

  /// Cached variation keys (flagKey → variationKey) for event attribution.
  Map<String, String>? _cachedDecisions;

  /// Fire-and-forget exposure tracking.
  ///
  /// Skipped for `'targeted'` rules — those are simple feature-flag rollouts
  /// with no experiment to attribute, so exposures would just be noise.
  void _trackExposure(Decision decision) {
    if (decision.ruleType == RuleType.targeted) return;

    final event = SignaKitEvent(
      eventKey: r'$exposure',
      userId: _truncate(userId, kMaxUserIdLength),
      timestamp: DateTime.now().toUtc().toIso8601String(),
      decisions: <String, String>{decision.flagKey: decision.variationKey},
      metadata: <String, Object?>{
        'flagKey': decision.flagKey,
        'variationKey': decision.variationKey,
        'ruleKey': decision.ruleKey,
      },
      attributes: _sanitizeAttributes(attributes),
    );

    // Fire-and-forget: errors are silently ignored.
    // ignore: discarded_futures
    _client.sendEventInternal(event).catchError((Object _) {});
  }

  /// Evaluate a single flag for this user.
  ///
  /// Returns `null` if the flag is not found or archived.
  Decision? decide(String flagKey) {
    if (_isBot) {
      return Decision(
        flagKey: flagKey,
        variationKey: 'off',
        enabled: false,
        ruleKey: null,
        ruleType: null,
        variables: const <String, VariableValue>{},
      );
    }

    final decision = _client.evaluateFlagInternal(flagKey, userId, attributes);
    if (decision != null) {
      (_cachedDecisions ??= <String, String>{})[flagKey] = decision.variationKey;
      _trackExposure(decision);
    }
    return decision;
  }

  /// Evaluate all flags for this user.
  Decisions decideAll() {
    if (_isBot) return _client.getBotDecisionsInternal();

    final decisions = _client.evaluateAllFlagsInternal(userId, attributes);

    final cache = <String, String>{};
    for (final entry in decisions.entries) {
      cache[entry.key] = entry.value.variationKey;
      _trackExposure(entry.value);
    }
    _cachedDecisions = cache;
    return decisions;
  }

  /// Track a conversion event for this user.
  Future<void> trackEvent(
    String eventKey, {
    num? value,
    Map<String, Object?>? metadata,
  }) async {
    if (_isBot) return;

    final sEventKey = _truncate(eventKey, kMaxEventKeyLength);
    final sUserId = _truncate(userId, kMaxUserIdLength);

    Map<String, Object?>? acceptedMetadata;
    if (metadata != null) {
      final encoded = jsonEncode(metadata);
      if (encoded.length <= kMaxMetadataSizeBytes) {
        acceptedMetadata = metadata;
      }
      // else: silently dropped (mirrors TS warn-and-drop)
    }

    final event = SignaKitEvent(
      eventKey: sEventKey,
      userId: sUserId,
      timestamp: DateTime.now().toUtc().toIso8601String(),
      attributes: _sanitizeAttributes(attributes),
      decisions: (_cachedDecisions != null && _cachedDecisions!.isNotEmpty)
          ? Map<String, String>.from(_cachedDecisions!)
          : null,
      value: value,
      metadata: acceptedMetadata,
    );

    await _client.sendEventInternal(event);
  }
}

File: packages/flags-golang/signakit/user_context.go
package signakit

import (
	"encoding/json"

	"github.com/signakit/flags-golang/internal/botua"
)

// isBotUA is a thin wrapper so client.go does not need to import the internal
// package directly.
func isBotUA(ua string) bool { return botua.IsBot(ua) }

// UserContext represents a single user being evaluated against the flag set.
// Not safe for concurrent use; create one per request/user.
type UserContext struct {
	client          *Client
	userID          string
	attributes      UserAttributes
	cachedDecisions map[string]string
	isBot           bool
}

// UserID returns the user identifier this context was created with.
func (u *UserContext) UserID() string { return u.userID }

// Attributes returns a copy-free view of the user's attributes (with the
// $userAgent key already stripped).
func (u *UserContext) Attributes() UserAttributes { return u.attributes }

// Decide evaluates a single flag for the user and returns the resulting
// Decision. If the flag is missing or archived, it returns nil.
//
// As a side effect, an "$exposure" event is dispatched (fire-and-forget by
// default) unless the decision came from a "targeted" rule.
func (u *UserContext) Decide(flagKey string) *Decision {
	if u.isBot {
		return &Decision{
			FlagKey:      flagKey,
			VariationKey: "off",
			Enabled:      false,
			RuleKey:      nil,
			RuleType:     nil,
			Variables:    map[string]any{},
		}
	}

	d := u.client.evaluateFlag(flagKey, u.userID, u.attributes)
	if d == nil {
		return nil
	}

	if u.cachedDecisions == nil {
		u.cachedDecisions = map[string]string{}
	}
	u.cachedDecisions[flagKey] = d.VariationKey

	u.trackExposure(*d)
	return d
}

// DecideAll evaluates every active flag for the user, returning a map of
// flagKey → Decision. Archived flags are omitted.
//
// Exposure events are dispatched for each non-targeted decision.
func (u *UserContext) DecideAll() Decisions {
	if u.isBot {
		return u.client.botDecisions()
	}

	decisions := u.client.evaluateAllFlags(u.userID, u.attributes)
	if u.cachedDecisions == nil {
		u.cachedDecisions = make(map[string]string, len(decisions))
	}
	for k, d := range decisions {
		u.cachedDecisions[k] = d.VariationKey
		u.trackExposure(d)
	}
	return decisions
}

// trackExposure dispatches an "$exposure" event, skipping decisions that came
// from a "targeted" rule (those are simple feature-flag rollouts and produce
// only noise on the exposures pipeline).
func (u *UserContext) trackExposure(d Decision) {
	if d.RuleType != nil && *d.RuleType == RuleTypeTargeted {
		return
	}
	event := Event{
		EventKey:  "$exposure",
		UserID:    truncate(u.userID, MaxUserIDLength),
		Timestamp: nowRFC3339(),
		Decisions: map[string]string{d.FlagKey: d.VariationKey},
		Metadata: map[string]any{
			"flagKey":      d.FlagKey,
			"variationKey": d.VariationKey,
			"ruleKey":      d.RuleKey, // *string serialises as null when nil
		},
	}
	if attrs := sanitizeAttributes(u.attributes); attrs != nil {
		event.Attributes = attrs
	}
	u.client.dispatchEvent(event)
}

// EventOption configures a TrackEvent call.
type EventOption func(*eventOptions)

type eventOptions struct {
	value    *float64
	metadata map[string]any
}

// WithValue attaches a numeric value (e.g. revenue) to the event.
func WithValue(v float64) EventOption {
	return func(o *eventOptions) { o.value = &v }
}

// WithMetadata attaches arbitrary metadata to the event. The map is dropped
// (with a logged warning) if it serialises to more than MaxMetadataSizeBytes.
func WithMetadata(m map[string]any) EventOption {
	return func(o *eventOptions) { o.metadata = m }
}

// TrackEvent records a conversion event for the user.
//
// For bots, this is a no-op. The event is dispatched fire-and-forget by
// default (see WithSyncEventDispatch for tests).
func (u *UserContext) TrackEvent(eventKey string, opts ...EventOption) {
	if u.isBot {
		return
	}

	cfg := &eventOptions{}
	for _, opt := range opts {
		opt(cfg)
	}

	event := Event{
		EventKey:  truncate(eventKey, MaxEventKeyLength),
		UserID:    truncate(u.userID, MaxUserIDLength),
		Timestamp: nowRFC3339(),
	}
	if attrs := sanitizeAttributes(u.attributes); attrs != nil {
		event.Attributes = attrs
	}
	if len(u.cachedDecisions) > 0 {
		event.Decisions = u.cachedDecisions
	}
	if cfg.value != nil {
		event.Value = cfg.value
	}
	if cfg.metadata != nil {
		raw, err := json.Marshal(cfg.metadata)
		switch {
		case err != nil:
			u.client.logger.Warn("signakit: metadata marshal failed; dropping", "err", err)
		case len(raw) > MaxMetadataSizeBytes:
			u.client.logger.Warn("signakit: metadata exceeds limit; dropping",
				"limit", MaxMetadataSizeBytes, "size", len(raw))
		default:
			event.Metadata = cfg.metadata
		}
	}

	u.client.dispatchEvent(event)
}

// sanitizeAttributes truncates keys/values and caps the total attribute count
// to match the limits defined in constants.go.
func sanitizeAttributes(attrs UserAttributes) UserAttributes {
	if len(attrs) == 0 {
		return nil
	}
	out := make(UserAttributes, len(attrs))
	count := 0
	for k, v := range attrs {
		if count >= MaxAttributesCount {
			break
		}
		count++
		key := truncate(k, MaxAttributeKeyLength)
		switch val := v.(type) {
		case nil:
			continue
		case string:
			out[key] = truncate(val, MaxAttributeValueLength)
		case []string:
			limit := len(val)
			if limit > 100 {
				limit = 100
			}
			truncated := make([]string, limit)
			for i := 0; i < limit; i++ {
				truncated[i] = truncate(val[i], MaxAttributeValueLength)
			}
			out[key] = truncated
		default:
			out[key] = val
		}
	}
	if len(out) == 0 {
		return nil
	}
	return out
}
